extra/exercism/testing/testing.factor

```diff
@@ -0,0 +1,227 @@
+M: dev-env config-exclusive?
+  sets:intersect { } = ;
+
+M: user-env config-exclusive?
+  \ config-exclusive? not-dev-env ;
+
+
+HOOK: config-matches-fs? project-env ( dirs problems deprecated -- ? )
+M: dev-env config-matches-fs?
+  [ over ] dip sets:intersect { } = -rot
+  [ natural-sort ] bi@ = and ;
+
+M: user-env config-matches-fs?
+  \ config-matches-fs? not-dev-env ;
+
+: (run-exercism-test) ( exercise -- )
+  dup "hello-world" = [ handle-hello-world ] when
+  [ "\ntesting exercise: %s\n\n" printf ]
+  [ exercise>filenames ]
+  bi
+  run-file run-test-file ;
+
+: wd-git-name ( -- name )
+  "git rev-parse --show-toplevel" utf8 [ contents ] with-process-reader*
+  nip 0 =
+  [ path-components last dup length 1 - head ]
+  [ drop "" ]
+  if ;
+
+PRIVATE>
+
+
+HOOK: verify-config project-env ( -- )
+M: dev-env verify-config
+  get-config-data dup problems>> [ deprecated>> ] dip 2dup
+  [ config-exclusive? ] 2dip
+
+  swap exercises-folder child-directories -rot
+  config-matches-fs?
+  and
+
+  exercises-folder child-directories
+  [ exercise>filenames [ exists? ] bi@ and ] all?
+  and
+
+  [ "config.json and exercises OK" print ]
+  [ "invalid config.json\n"
+    print 2 exit ]
+  if ;
+
+M: user-env verify-config
+  exercises-folder child-directories
+  [ exercise>filenames [ exists? ] bi@ and ] all?
+
+  [ "config OK: all problems have implementations and unit tests" print ]
+  [ "invalid config: problems are missing implementations or tests\n"
+    print 2 exit ]
+  if ;
+
+M: f verify-config
+  \ verify-config not-an-exercism-folder ;
+
+
+HOOK: run-exercism-test project-env ( exercise -- )
+M: dev-env run-exercism-test
+  (run-exercism-test) ;
+
+M: user-env run-exercism-test
+  (run-exercism-test) ;
+
+M: f run-exercism-test
+  drop \ run-exercism-test not-an-exercism-folder ;
+
+
+: run-all-exercism-tests ( -- )
+  exercises-folder child-directories [ run-exercism-test ] each ;
+
+: choose-suite ( arg -- )
+  {
+    { [ dup "VERIFY"  =      ] [ drop verify-config ] }
+    { [ dup "run-all" =      ] [ drop verify-config run-all-exercism-tests ] }
+    { [ dup exercise-exists? ] [ verify-config run-exercism-test ] }
+      [ verify-config "exercism.testing: choose-suite: bad last argument `%s', expected 'run-all' or an exercise slug\n\n" printf ]
+  } cond ;
+
+: guess-project-env ( -- )
+  "exercises" { ".keep" "hello-world" }
+  [ append-path ] with map
+  {
+    "config.json"
+    ".git"
+    ".gitignore"
+    "exercises"
+  }
+  append
+  [ exists? ] all?
+  "xfactor"
+  [ wd-git-name = ]
+  [ ".." prepend-path absolute-path current-directory namespaces:get = ]
+  bi and and dup [ T{ dev-env } project-env namespaces:set ] when
+
+  {
+    "hello-world"
+  } [ exists? ] all?
+  dup [ T{ user-env } project-env namespaces:set ] when
+
+  xor
+  [
+    current-directory project-env [ namespaces:get ] bi@
+    "working directory OK: %s is a %s \n" printf
+  ]
+  [
+    current-directory namespaces:get
```

Comment:
You shouldn't use ``current-directory set``, it's bad practice because it doesn't restore the previous value when the computation is completed, thus losing it.  Better would be ``path current-directory [ ... ] with-variable``